File: _BHCVRP_Python_Version/cujae_inf_citi_om/data/TimeWindow.py
from exceptions.ServiceTimeException import ServiceTimeException
from exceptions.TimeWindowException import TimeWindowException

# Clase que modela una ventana de tiempo para la variante VRPTW

class TimeWindow:
    def __init__(self, initial_node=None, end_node=None, service_time=None):
        if (
            initial_node is not None
            and end_node is not None
            and service_time is not None
        ):
            # Constructor con tres argumentos
            self._initial_node = initial_node
            self._end_node = end_node
            self._service_time = service_time

        else:
            # Constructor sin argumentos (o con argumentos predeterminados)
            self._initial_node = 0.0
            self._end_node = 0.0
            self._service_time = 0.0

    def get_initial_node(self):
        return self._initial_node

    def set_initial_node(self, initial_node):
        if initial_node >= 0:
            self._initial_node = initial_node
        else:
            raise TimeWindowException("La ventana de tiempo no puede ser negativa")

    def get_end_node(self):
        return self._end_node

    def set_end_node(self, end_node):
        if end_node >= 0:
            self._end_node = end_node
        else:
            raise TimeWindowException("La ventana de tiempo no puede ser negativa")

    def get_service_time(self):
        return self._service_time

    def set_service_time(self, service_time):
        if service_time >= 0:
            self._service_time = service_time
        else:
            raise ServiceTimeException("El tiempo de servicio no puede ser negativo")

File: BHCVRP_Java_Version/src/cujae/inf/citi/om/data/TimeWindow.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package cujae.inf.citi.om.data;

import cujae.inf.citi.om.exceptions.ServiceTimeException;
import cujae.inf.citi.om.exceptions.TimeWindowException;

/**
 *
 * @author kmych
 */
public class TimeWindow {
    private double initialNode;
    private double endNode;
    private double serviceTime;
    
    public TimeWindow(){
        super();
    }
    
    public TimeWindow(float initialNode, float endNode, float serviceTime){
        
    }
    
    public double getInitialNode(){
        return initialNode;
    }
    
    public void setInitialNode(double initialNode) throws TimeWindowException{
        if(initialNode >= 0){
            this.initialNode = initialNode;
        }
        else
            throw new TimeWindowException("El nodo inicial debe ser mayor que cero.");   
    }
    
    public double getEndNode(){
        return initialNode;
    }
    
    public void setEndNode(double endNode) throws TimeWindowException{
        if(endNode >= 0){
            this.endNode = endNode;
        }
        else
            throw new TimeWindowException("El nodo final debe ser mayor que cero.");     
    }
    
    public double getServiceTime(){
        return initialNode;
    }
    
    public void setServiceTime(double serviceTime) throws ServiceTimeException{
        if(serviceTime >= 0){
            this.serviceTime = serviceTime;
        }
        else
            throw new ServiceTimeException("El tiempo de servicio tiene que ser mayor que cero.");
    }
    
    public double calculateExitTime(double travelTime, double arrivalTime){
        return 0; //max(travel_time, arrive_time) + serviceTime
    }
    
    public boolean verifyTW(double arriveTime){
        return arriveTime <= this.endNode;
    }
}
